login-service/src/model/postgres/AuthClient.entity.ts

```diff
@@ -0,0 +1,167 @@
+     */
+    @Column()
+    requiresSecret: boolean;
+
+    /**
+     * A list of all login events that this client caused.
+     */
+    @OneToMany(() => ActiveLogin, (login) => login.createdByClient)
+    @ApiHideProperty()
+    loginsOfThisClient: Promise<ActiveLogin[]>;
+
+    /**
+     * Calculated the sha256 hash of the input.
+     *
+     * Can be used as fingerprint for the secret hashes.
+     * @param hash The hash or other data to calculate the fingerprint for
+     * @returns The hex string of the sha256 hash
+     */
+    private fingerprint(hash: string): string {
+        return crypto.createHash("sha256").update(hash).digest("hex");
+    }
+
+    /**
+     * Generates a new secret and adds it to the list of this client.
+     * Does NOT save the entitiy!
+     *
+     * **Note**: The secret text is only returned here and will not be saved as plain text.
+     * There is no way to retrieve it later.
+     * Only a hashed version and a 5 letter prefix (for easier identification) will be saved
+     *
+     * The generated secret will be hex encoded random bytes of length GROPIUS_CLIENT_SECRET_LENGTH
+     * @returns The generated secret text, the fingerprint of the hash of the secret and the censored version
+     */
+    async addSecret(): Promise<{
+        secretText: string;
+        fingerprint: string;
+        censored: string;
+    }> {
+        const length = Math.min(15, parseInt(process.env.GROPIUS_CLIENT_SECRET_LENGTH, 10));
+        const secretText = (await randomBytesAsync(length)).toString("hex");
+        if (secretText.length < 15) {
+            throw new Error("Secret must be at least 15 characters long");
+        }
+        if (secretText.match(/[^a-zA-Z0-9+/-_=]/)) {
+            throw new Error("Secret can not match /[^a-zA-Z0-9+/-_=]/");
+        }
+        const hash = await bcrypt.hash(secretText, parseInt(process.env.GROPIUS_BCRYPT_HASH_ROUNDS, 10));
+        const censored = secretText.substring(0, 5);
+        if (!this.clientSecrets?.length || !this.clientSecrets?.push) {
+            this.clientSecrets = [];
+        }
+        this.clientSecrets.push(censored + ";" + hash);
+
+        return {
+            secretText,
+            fingerprint: this.fingerprint(hash),
+            censored: censored + "**********",
+        };
+    }
+
+    /**
+     * Returns a list containing not only the string stored in the database but also the 5 letter censored prefix and the fingerprint of the hash.
+     *
+     * The result of this should NOT be exposed as it contains the full hash.
+     * @returns A list of the full stored hash, censore version and the fingerprint of the hash for every secret
+     */
+    getFullHashesPlusCensoredAndFingerprint(): {
+        secret: string;
+        censored: string;
+        fingerprint: string;
+    }[] {
+        return this.clientSecrets.map((s) => {
+            const semiIndex = s.indexOf(";");
+            const censored = s.substring(0, semiIndex) + "**********";
```

Comment:
The number of starts is fixed to _below_ what it could ever be on purpose to reveal less information (yes I know the secret text length is fixed, but it might change sometime)